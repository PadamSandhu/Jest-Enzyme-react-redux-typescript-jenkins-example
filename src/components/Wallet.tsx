import * as React from "react";
import { connect } from "react-redux";
import { deposit, withdraw } from "../store/actions";
import { bindActionCreators } from "redux";

export class Wallet extends React.Component<any, any> {
    constructor(props) {
        super(props);
        this.state = { balance: undefined };
    }

    updateBalance = (event) => this.setState({balance: parseInt(event.target.value, 10)});
    deposits = () => this.props.deposit( this.state.balance );
    withdrawl = () => this.props.withdraw( this.state.balance );

    render() {
        return (
            <div>
                <h2 className="balance">Wallet balance: { this.props.balance }</h2>
                <br/>
                <input type="number" className="input-wallet" onChange = { this.updateBalance}/>
                <button className="btn-deposit" onClick = { this.deposits }> Deposit </button>
                <button className="btn-withdrawl" onClick = { this.withdrawl }> Withdraw!!! </button>
            </div>
        );
    }
}

function mapStateToProps(state) {
    return { balance: state.balance };
}
function mapDispatchToProps(dispath) {
    return bindActionCreators({ deposit , withdraw }, dispath);
}

// A compact way of creating mapStateToProps is used in Loot.tsx Example
export default connect( mapStateToProps, mapDispatchToProps )(Wallet);   // Whole state in redux is just a number for now !!
