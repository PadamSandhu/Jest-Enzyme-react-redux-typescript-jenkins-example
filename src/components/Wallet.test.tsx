import * as React from "react";
import { shallow } from "enzyme";
import { Wallet } from "./Wallet";   // Can't Add Redux-store component !!

describe("Wallet", () => {
    const mockDeposit = jest.fn();    // This is the mock Function which is required to simulate an action !!
    const mockWithdraw = jest.fn();
    const props = { balance: 20, deposit: mockDeposit, withdraw: mockWithdraw };    // Step 2 for action, provide a key in the props!!
    const wallet = shallow(<Wallet {...props} />);
    it("renders properly", () => {
        expect(wallet).toMatchSnapshot();
    });

    // Check if the balance is available as props !!
    it("displays the balance from props", () => {
        expect(wallet.find(".balance").text()).toEqual("Wallet balance: 20");
    });

    it("Creates an input to deposit or withdraw into balance", () => {
        expect(wallet.find(".input-wallet").exists()).toBe(true);
    });

    describe("when the user types into the wallet input", () => {
        const userBalance = "25";

        beforeEach(() => {
            wallet.find(".input-wallet").simulate("change" , { target: {value: userBalance}});
        });

        it("Updates the local balance to the `state` and converts the number into string", () => {
            expect(wallet.state().balance).toEqual(parseInt(userBalance, 10));
        });


        describe("The user wants to make a deposit", () => {
            beforeEach(() => wallet.find(".btn-deposit").simulate("click"));

            it("Dispathces the `deposit()` it recives from props with local balance", () => {
                expect(mockDeposit).toHaveBeenCalledWith(parseInt(userBalance, 10));
            });
        });

        describe("The user wants to make a WithDrawl", () => {
            beforeEach(() => wallet.find(".btn-withdrawl").simulate("click"));

            it("Dispathces the `withdraw()` it recives from props with local balance", () => {
                expect(mockWithdraw).toHaveBeenCalledWith(parseInt(userBalance, 10));
            });
        });

    });


});
