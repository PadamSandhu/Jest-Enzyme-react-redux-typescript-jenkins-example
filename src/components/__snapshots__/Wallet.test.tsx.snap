// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wallet renders properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Wallet
    balance={20}
    deposit={[MockFunction]}
    withdraw={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2
          className="balance"
        >
          Wallet balance: 
          20
        </h2>,
        <br />,
        <input
          className="input-wallet"
          onChange={[Function]}
          type="number"
        />,
        <button
          className="btn-deposit"
          onClick={[Function]}
        >
           Deposit 
        </button>,
        <button
          className="btn-withdrawl"
          onClick={[Function]}
        >
           Withdraw!!! 
        </button>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Wallet balance: ",
            20,
          ],
          "className": "balance",
        },
        "ref": null,
        "rendered": Array [
          "Wallet balance: ",
          20,
        ],
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "br",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "input-wallet",
          "onChange": [Function],
          "type": "number",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": " Deposit ",
          "className": "btn-deposit",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": " Deposit ",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": " Withdraw!!! ",
          "className": "btn-withdrawl",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": " Withdraw!!! ",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2
            className="balance"
          >
            Wallet balance: 
            20
          </h2>,
          <br />,
          <input
            className="input-wallet"
            onChange={[Function]}
            type="number"
          />,
          <button
            className="btn-deposit"
            onClick={[Function]}
          >
             Deposit 
          </button>,
          <button
            className="btn-withdrawl"
            onClick={[Function]}
          >
             Withdraw!!! 
          </button>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Wallet balance: ",
              20,
            ],
            "className": "balance",
          },
          "ref": null,
          "rendered": Array [
            "Wallet balance: ",
            20,
          ],
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "input-wallet",
            "onChange": [Function],
            "type": "number",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": " Deposit ",
            "className": "btn-deposit",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": " Deposit ",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": " Withdraw!!! ",
            "className": "btn-withdrawl",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": " Withdraw!!! ",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
